File: apps/ios/Core/ConversationActivity.swift
import Foundation

/// An unanswered check-in never extends the paid session. Times use a monotonic clock.
public struct ConversationActivity: Sendable {
    public enum Action: Equatable, Sendable { case wait, checkIn, warning(Int), end }
    public static let quietSeconds: Double = SessionLimits.idleVoiceSeconds
    public static let checkInSeconds: Double = 15
    public static let speechGraceSeconds: Double = 15
    public static let typingGraceSeconds: Double = 60
    public static let responseGraceSeconds: Double = 45
    private var quietSince: Double
    private var outputDeadline: Double
    private var lastInput: Double?
    private var typingStarted: Double?
    private var lastTyping: Double?
    private var busyStarted: Double?
    private var checkedIn = false
    public init(now: Double) { quietSince = now; outputDeadline = now + 60 }
    public mutating func learnerEngaged(now: Double) {
        quietSince = now; outputDeadline = now + 60; checkedIn = false
        typingStarted = nil; lastTyping = nil; lastInput = nil; busyStarted = nil
    }
    public mutating func assistantActive(now: Double) {
        if !checkedIn && now <= outputDeadline { quietSince = now }
    }
    public mutating func inputActive(now: Double) { lastInput = now }
    public mutating func typing(now: Double) {
        if typingStarted == nil { typingStarted = now }
        lastTyping = now
    }
    public mutating func tick(now: Double, muted: Bool = false, busy: Bool = false) -> Action {
        if busy && busyStarted == nil { busyStarted = now }
        if !busy { busyStarted = nil }
        let recentInput = !muted && lastInput.map { now - $0 < 1.5 } == true
        let editing = lastTyping.map { now - $0 < 10 } == true
        var deadline = quietSince + Self.quietSeconds
        // Audio levels provide bounded protection for delayed transcripts, never unlimited activity.
        if recentInput { deadline += Self.speechGraceSeconds }
        if editing, let started = typingStarted { deadline = max(deadline, started + Self.typingGraceSeconds) }
        if busy, let started = busyStarted { deadline = max(deadline, started + Self.responseGraceSeconds) }
        if now >= deadline { return .end }
        if deadline - now <= 5 { return .warning(Int(ceil(deadline - now))) }
        if !checkedIn && !muted && !recentInput && !editing && !busy && now - quietSince >= Self.checkInSeconds {
            checkedIn = true; return .checkIn
        }
        return .wait
    }
}

File: apps/android/app/src/main/java/chat/mural/core/ConversationActivity.kt
package chat.mural.core

import kotlin.math.ceil
import kotlin.math.max

/** An unanswered check-in never extends the paid session. Times use a monotonic clock. */
class ConversationActivity(now: Double) {
    sealed interface Action {
        data object Wait : Action
        data object CheckIn : Action
        data class Warning(val seconds: Int) : Action
        data object End : Action
    }
    private var quietSince = now
    private var outputDeadline = now + 60
    private var lastInput: Double? = null
    private var typingStarted: Double? = null
    private var lastTyping: Double? = null
    private var busyStarted: Double? = null
    private var checkedIn = false
    fun learnerEngaged(now: Double) {
        quietSince = now; outputDeadline = now + 60; checkedIn = false
        typingStarted = null; lastTyping = null; lastInput = null; busyStarted = null
    }
    fun assistantActive(now: Double) { if (!checkedIn && now <= outputDeadline) quietSince = now }
    fun inputActive(now: Double) { lastInput = now }
    fun typing(now: Double) { if (typingStarted == null) typingStarted = now; lastTyping = now }
    fun tick(now: Double, muted: Boolean = false, busy: Boolean = false): Action {
        if (busy && busyStarted == null) busyStarted = now
        if (!busy) busyStarted = null
        val recentInput = !muted && lastInput?.let { now - it < 1.5 } == true
        val editing = lastTyping?.let { now - it < 10 } == true
        var deadline = quietSince + QUIET_SECONDS
        if (recentInput) deadline += SPEECH_GRACE_SECONDS
        if (editing) typingStarted?.let { deadline = max(deadline, it + TYPING_GRACE_SECONDS) }
        if (busy) busyStarted?.let { deadline = max(deadline, it + RESPONSE_GRACE_SECONDS) }
        if (now >= deadline) return Action.End
        if (deadline - now <= 5) return Action.Warning(ceil(deadline - now).toInt())
        if (!checkedIn && !muted && !recentInput && !editing && !busy && now - quietSince >= CHECK_IN_SECONDS) {
            checkedIn = true; return Action.CheckIn
        }
        return Action.Wait
    }
    companion object {
        const val QUIET_SECONDS = SessionLimits.IDLE_VOICE_SECONDS
        const val CHECK_IN_SECONDS = 15.0
        const val SPEECH_GRACE_SECONDS = 15.0
        const val TYPING_GRACE_SECONDS = 60.0
        const val RESPONSE_GRACE_SECONDS = 45.0
    }
}
